Support size oblivious max equation

bobrenjc93 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #147344

Addresses https://github.com/pytorch/pytorch/issues/125914 by detecting when we have a sym_max between {0, 1} and a summation of size-like unbacked symints.

The basic idea is max(1, u0 + u1) can be simplified to u0 + u1 if both u0 and u1 are size-like since their value ranges are [2, inf].

cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @chenyang78 @kadeng @chauhang @amjames

bobrenjc93 committed:
Support size oblivious max equation

[ghstack-poisoned]

bobrenjc93 committed:
Update on "Support size oblivious max equation"


Partially addresses #https://github.com/pytorch/pytorch/issues/125914

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bobrenjc93 commented:
@pytorchbot rebase -b main

bobrenjc93 committed:
Update on "Support size oblivious max equation"


Partially addresses #https://github.com/pytorch/pytorch/issues/125914

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

angelayi commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -5771,6 +5772,25 @@ def simplify(self, expr: _SympyT) -> _SympyT:
+                    if (
+                        isinstance(b, Add)
+                        and len(b.free_symbols) == 2  # TODO: expand to N?
+                        and b.free_symbols == set(b.atoms())
```
Could this be expanded to work for Mul ops, or if there contains other integers that are greater than 0? Or is that too complicated?

pianpwk commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -5771,6 +5772,25 @@ def simplify(self, expr: _SympyT) -> _SympyT:
+            if max_replacements:
+                expr = expr.xreplace(max_replacements)
+                expr = safe_expand(expr)
+
```
shouldn't we only be doing this under size-oblivious reasoning?

Also, I wonder if we can do this with ValueRange reasoning, where if we find that we have `Max(x, b)` where `x < bound_sympy(b).lower`, or `Min(y, b)` where `y > bound_sympy(b).upper`, we can replace with b.

laithsakka commented on test/dynamo/test_misc.py:
```diff
@@ -10859,6 +10859,21 @@ def fn(x, y):
+            torch._check_is_size(u0)
+            torch._check_is_size(u1)
+            torch._check(u0 + u1 == 20)
+            if guard_size_oblivious(torch.sym_max(1, u0 + u1) == 20):
```
can you update the summary to say something around if we have torch.sym_max(1/0, u0 + u1) inside
guard_size_oblivious then this diff simplify that to torch.sym_max(1/0, u0 + u1) because u0+u1 are known to be >=1 each hence

laithsakka commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -5771,6 +5772,25 @@ def simplify(self, expr: _SympyT) -> _SympyT:
+        if expr.has(Max):
+            max_replacements = {}
+            for atom in expr.atoms(Max):
+                a, b = atom.args
```
i mean what about
torch._check(u0 + u1 + u2 + u3+ u4 + u5 == 20)
if guard_size_oblivious(torch.sym_max(6, u0 + u1 + u2 + u3+ u4 + u5+..) == 20):

this feels like a fix for a specific use case, I do am not sure how i feel about it.

laithsakka commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -5771,6 +5772,25 @@ def simplify(self, expr: _SympyT) -> _SympyT:
+            max_replacements = {}
+            for atom in expr.atoms(Max):
+                a, b = atom.args
+                if b == 1 or b == 0:
```
a more genera solution could be
for max(a, b)
find the lower a and of interval a and lower bound for interval b.
if one is smaller than the other resolve it .

bobrenjc93 replied:
Yeah I agree, I think the tricky part is balancing the computational complexity with how realistic the case is. My current stance is to start with cases that real users ran into (eg. this one) and then slowly increase the complexity of analysis from there. I'll put up another set of diffs that addresses 1) the N addition case (eg. u0 + u1 + ... un) 2) the min cases and 3) the mul case.

laithsakka replied:
mmm If the user have another way around it with torch.check, we should not support it this way. because then we will end up with so many patches like this that we have to support backward compatibility for.

and we end up in a situation were we cant explain how does the reasoning work, and if we do i till be a long list of 
if and if and if then ,,

bobrenjc93 replied:
So taking a step back, I'm a believer that the interface should drive the implementation and not the other way around.

We've made a concrete contract with the user that for unbacked reasoning, users can assume if a symbol is size-like, that we'll correctly do the inference and simplification for cases like this. I view the inability to adhere to this user contract a bug. 

torch._check is an API that gives user the affordance to tell us things that we can't assume. For example if a user knows for sure an input won't be larger than a certain size. This example is one where we should be able to statically solve it given the contract we've laid out and thus I don't think we should expect users to have to explicitly torch._check this. 

I think it's much better to make incremental progress to increase coverage to our provided contract than to limit progress due to fears that users will actually follow through with our contract. If we don't believe the contract is right, then that's a completely different discussion that should be held in a different forum.

laithsakka commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -5771,6 +5772,25 @@ def simplify(self, expr: _SympyT) -> _SympyT:
+                    a, b = b, a
+                if a == 1 or a == 0:
+                    if (
+                        isinstance(b, Add)
```
yeh very customized on both the 0/1 up there and the a+b here. 
the issue with landing changes like this, is that if users saw this working then will assume it generalize but it wont. its kind of a branchy behaviour. 

is the user allowed to do something like
torch._check(torch.sym_max(1, u0 + u1) == 20) to go around this issue

bobrenjc93 committed:
Update on "Support size oblivious max equation"


Addresses https://github.com/pytorch/pytorch/issues/125914 by detecting when we have a sym_max between {0, 1} and a summation of size-like unbacked symints.

The basic idea is max(1, u0 + u1) can be simplified to u0 + u1 if both u0 and u1 are size-like since their value ranges are [2, inf].

[ghstack-poisoned]

bobrenjc93 commented:
@pytorchbot merge
